### PR TITLE
Tweak how source paths are passed to the compiler

### LINKED ISSUE #209: Paths printed relative to the Cargo.toml file interfere with M-x compile
If you're in a subdirectory of Cargo repository (e.g. `src`) and run `cargo build`, `rustc` is run from the directory with `Cargo.toml`, meaning all printed paths are relevative to that directory, *not* the one where `cargo build` was run.

Emac's M-x compile mode will "hyperlink" paths in error messages to allow jumping straight to the error (I'm sure other editors provide this functionality too), but it is taken relative to the folder of the buffer on which `M-x compile` was run (i.e. if you are editing `.../src/lib.rs` and run `M-x compile cargo build` then emacs assumes the diagnostics are relative to `.../src`). Hence, the automatic running of `rustc` in a different directory makes it harder to use that mode.

This would theoretically be fixed by running the `rustc` subprocess from the directory in which `cargo` is invoked.

### LINKED ISSUE #694: Cargo prints warnings with full paths
`cargo build` prints compile time warnings using full file paths. This results in very long lines making it either wrap on a moderately widened terminal window and/or hard to follow.

Sample cargo output:
```
$ cargo build
   Compiling hello_world v0.0.1 (file:///Users/kostaskarachalios/Code/Learning/Rust/guide/hello_world)
/Users/kostaskarachalios/Code/Learning/Rust/guide/hello_world/src/main.rs:2:9: 2:14 warning: value assigned to `x` is never read, #[warn(dead_assignment)] on by default
/Users/kostaskarachalios/Code/Learning/Rust/guide/hello_world/src/main.rs:2     let mut x = 5i;
                                                                                    ^~~~~
```

For comparison:
```
$ rustc src/main.rs
src/main.rs:2:9: 2:14 warning: value assigned to `x` is never read, #[warn(dead_assignment)] on by default
src/main.rs:2     let mut x = 5i;
                      ^~~~~
```

Versions: 
```
$ cargo -V
cargo 0.0.1-pre-nightly (0881adf 2014-10-10 22:03:11 -0700)
$ rustc -v
rustc 0.13.0-nightly (adb44f53d 2014-10-12 00:07:15 +0000)
```

### LINKED ISSUE #1081: cargo doc generated files contain references to absolute pathes
```
<meta name="description" content="Source to the Rust file `src/lib.rs`.">
<meta name="description" content="Source to the Rust file `/home/user/…/src/lib.rs`.">
```

The first description as generated by invoking `rustdoc` directly. The second one, generated by invoking `cargo doc`, contains full path to the file on the developer machine.

### PR DESCRIPTION
**Pull request by alexcrichton**
All paths printed will now be absolute paths unless the path is a descendant of
the current directory. This should keep error messages and warnings of a
reasonable length when working with the local project while still allowing
errors in registry/git dependencies to be tracked down.

Special care is taken in these situations to ensure that the error message from
the compiler prints a reasonable path.

Closes #209
Closes #694
Closes #1081

### CODE REVIEW

File: src/cargo/ops/cargo_rustc/fingerprint.rs [hunk 1]
```diff
@@ -289,3 +296,18 @@ fn filename(target: &Target) -> String {
     };
     format!("{}{}-{}", flavor, kind, target.get_name())
 }
+
+// The dep-info files emitted by the compiler all have their listed paths
```
Comment (huonw) [inline src/cargo/ops/cargo_rustc/fingerprint.rs]: Eurgh. Shouldn't these be absolute?

Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/fingerprint.rs]: Possibly yeah, the source of the "problem" is that we call `rustc src/foo.rs` so the deps look like `src/foo.rs` and `src/bar.rs` instead of `/path/to/src/foo.rs`. I think that this mirrors gcc behavior, however. Either way, it's pretty easy to tweak in the future.
